# ETL Pipeline Preparation
Follow the instructions below to help you create your ETL pipeline.
### 1. Import libraries and load datasets.
- Import Python libraries
- Load `messages.csv` into a dataframe and inspect the first few lines.
- Load `categories.csv` into a dataframe and inspect the first few lines.

In [18]:
# import libraries
import pandas as pd 
import numpy as np 
from sqlalchemy import create_engine

In [19]:
# load messages dataset
messages = pd.read_csv('messages.csv')
messages.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  
0  Un front froid se retrouve sur Cuba ce matin. ...  direct  
1                 Cyclone nan fini osinon li pa fini  direct  
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct  
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct  
4  facade ouest d Haiti et le reste du pays aujou...  direct

In [20]:
# load categories dataset
categories = pd.read_csv('categories.csv')
categories.head()

id                                         categories
0   2  related-1;request-0;offer-0;aid_related-0;medi...
1   7  related-1;request-0;offer-0;aid_related-1;medi...
2   8  related-1;request-0;offer-0;aid_related-0;medi...
3   9  related-1;request-1;offer-0;aid_related-1;medi...
4  12  related-1;request-0;offer-0;aid_related-0;medi...

## Clean up first .. before merging
I will clean up the data and put in in the right format before merging them together. I found that to be easier and more logical. 

### 2 . Setting up the indexes 

In [21]:
messages=messages.set_index('id')
categories=categories.set_index('id')

### 3. Split `categories` into separate category columns.
- Split the values in the `categories` column on the `;` character so that each value becomes a separate column. You'll find [this method](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.Series.str.split.html) very helpful! Make sure to set `expand=True`.
- Use the first row of categories dataframe to create column names for the categories data.
- Rename columns of `categories` with new column names.

In [22]:
categories = categories['categories'].str.split(';', expand=True)


In [23]:
row = categories.iloc[1,:]
row_splitted= row.str.split('-', expand=True)
columns= row_splitted.iloc[:,0]
categories.columns = columns
categories.head()

0     related    request    offer    aid_related    medical_help  \
id                                                                 
2   related-1  request-0  offer-0  aid_related-0  medical_help-0   
7   related-1  request-0  offer-0  aid_related-1  medical_help-0   
8   related-1  request-0  offer-0  aid_related-0  medical_help-0   
9   related-1  request-1  offer-0  aid_related-1  medical_help-0   
12  related-1  request-0  offer-0  aid_related-0  medical_help-0   

0     medical_products    search_and_rescue    security    military  \
id                                                                    
2   medical_products-0  search_and_rescue-0  security-0  military-0   
7   medical_products-0  search_and_rescue-0  security-0  military-0   
8   medical_products-0  search_and_rescue-0  security-0  military-0   
9   medical_products-1  search_and_rescue-0  security-0  military-0   
12  medical_products-0  search_and_rescue-0  security-0  military-0   

0     child_alone       ...           aid_centers    other_infrastructure  \
id                      ...                                                 
2   child_alone-0       ...         aid_centers-0  other_infrastructure-0   
7   child_alone-0       ...         aid_centers-0  other_infrastructure-0   
8   child_alone-0       ...         aid_centers-0  other_infrastructure-0   
9   child_alone-0       ...         aid_centers-0  other_infrastructure-0   
12  child_alone-0       ...         aid_centers-0  other_infrastructure-0   

0     weather_related    floods    storm    fire    earthquake    cold  \
id                                                                       
2   weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
7   weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0   
8   weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
9   weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
12  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   

0     other_weather    direct_report  
id                                    
2   other_weather-0  direct_report-0  
7   other_weather-0  direct_report-0  
8   other_weather-0  direct_report-0  
9   other_weather-0  direct_report-0  
12  other_weather-0  direct_report-0  

[5 rows x 36 columns]

### 4. Convert category values to just numbers 0 or 1.
- Iterate through the category columns in df to keep only the last character of each string (the 1 or 0). For example, `related-0` becomes `0`, `related-1` becomes `1`. Convert the string to a numeric value.
- You can perform [normal string actions on Pandas Series](https://pandas.pydata.org/pandas-docs/stable/text.html#indexing-with-str), like indexing, by including `.str` after the Series. You may need to first convert the Series to be of type string, which you can do with `astype(str)`.

In [24]:
for name in columns: 
    categories[name] = categories[name].str[-1:]

In [25]:
categories.head()

0  related request offer aid_related medical_help medical_products  \
id                                                                   
2        1       0     0           0            0                0   
7        1       0     0           1            0                0   
8        1       0     0           0            0                0   
9        1       1     0           1            0                1   
12       1       0     0           0            0                0   

0  search_and_rescue security military child_alone      ...      aid_centers  \
id                                                      ...                    
2                  0        0        0           0      ...                0   
7                  0        0        0           0      ...                0   
8                  0        0        0           0      ...                0   
9                  0        0        0           0      ...                0   
12                 0        0        0           0      ...                0   

0  other_infrastructure weather_related floods storm fire earthquake cold  \
id                                                                          
2                     0               0      0     0    0          0    0   
7                     0               1      0     1    0          0    0   
8                     0               0      0     0    0          0    0   
9                     0               0      0     0    0          0    0   
12                    0               0      0     0    0          0    0   

0  other_weather direct_report  
id                              
2              0             0  
7              0             0  
8              0             0  
9              0             0  
12             0             0  

[5 rows x 36 columns]

#### Converting strings to integers

In [26]:
categories = categories.astype(int)

#### Second look at the data 
I am returning to this point because after I proceeded with the analysis, I found out that some values are not either 0 or 1. So I would like to check here.

In [27]:
categories.groupby(['related'])['request'].count()

related
0     6125
1    19930
2      193
Name: request, dtype: int64

It turns out, we do have non-binary data here. There are 193 records with value 2 in the `related` column. we will figure out the relationship between the value and the content of the message after the merge. 

### 5. merging the data 


In [28]:
# merge datasets
df = messages.join(categories, on='id')
df.head()

message  \
id                                                      
2   Weather update - a cold front from Cuba that c...   
7             Is the Hurricane over or is it not over   
8                     Looking for someone but no name   
9   UN reports Leogane 80-90 destroyed. Only Hospi...   
12  says: west side of Haiti, rest of the country ...   

                                             original   genre  related  \
id                                                                       
2   Un front froid se retrouve sur Cuba ce matin. ...  direct        1   
7                  Cyclone nan fini osinon li pa fini  direct        1   
8   Patnm, di Maryani relem pou li banm nouvel li ...  direct        1   
9   UN reports Leogane 80-90 destroyed. Only Hospi...  direct        1   
12  facade ouest d Haiti et le reste du pays aujou...  direct        1   

    request  offer  aid_related  medical_help  medical_products  \
id                                                                
2         0      0            0             0                 0   
7         0      0            1             0                 0   
8         0      0            0             0                 0   
9         1      0            1             0                 1   
12        0      0            0             0                 0   

    search_and_rescue      ...        aid_centers  other_infrastructure  \
id                         ...                                            
2                   0      ...                  0                     0   
7                   0      ...                  0                     0   
8                   0      ...                  0                     0   
9                   0      ...                  0                     0   
12                  0      ...                  0                     0   

    weather_related  floods  storm  fire  earthquake  cold  other_weather  \
id                                                                          
2                 0       0      0     0           0     0              0   
7                 1       0      1     0           0     0              0   
8                 0       0      0     0           0     0              0   
9                 0       0      0     0           0     0              0   
12                0       0      0     0           0     0              0   

    direct_report  
id                 
2               0  
7               0  
8               0  
9               0  
12              0  

[5 rows x 39 columns]

In [29]:
df[df.related == 2].message

id
146      Dans la zone de Saint Etienne la route de Jacm...
263      . .. i with limited means. Certain patients co...
373      The internet caf Net@le that's by the Dal road...
565      Bonsoir, on est a bon repos aprs la compagnie ...
700      URGENT CRECHE ORPHANAGE KAY TOUT TIMOUN CROIX ...
804      elle est vraiment malade et a besoin d'aide. u...
804      elle est vraiment malade et a besoin d'aide. u...
804      elle est vraiment malade et a besoin d'aide. u...
804      elle est vraiment malade et a besoin d'aide. u...
1063     no authority has passed by to see us. We don't...
1080     It's Over in Gressier. The population in the a...
1113     we sleep with the baby. Thanks in advance for ...
1121     I need help in Jrmie because I was in Port-au-...
1123     fsa pou mwen s v p map mouri mwen gen tout po ...
1460     .. Gonaives, in a place called Canal Bois in F...
1483     GEN YON TIBEBE KI MALAD NAN KOU PASKE BLOK TON...
1550     don't understand the first part. .. understa

Reading few of the messages above, it looks like most of them were not in english and hard to classify. I doubt that keeping them will add value to our analysis. Therefore, I will just drop them.  

In [30]:
df = df[df.related != 2]

### 6. Remove duplicates.
- Check how many duplicates are in this dataset.
- Drop the duplicates.
- Confirm duplicates were removed.

In [31]:
# check number of duplicates
df.duplicated().sum()

155

In [32]:
# drop duplicates
df.drop_duplicates(inplace =True)

In [33]:
# check number of duplicates
df.duplicated().sum()

0

### 7. Save the clean dataset into an sqlite database.
You can do this with pandas [`to_sql` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_sql.html) combined with the SQLAlchemy library. Remember to import SQLAlchemy's `create_engine` in the first cell of this notebook to use it below.

In [34]:
engine = create_engine('sqlite:///DB_messeges.db')
df.to_sql('T_messeges', engine, index=False)

### 8. Use this notebook to complete `etl_pipeline.py`
Use the template file attached in the Resources folder to write a script that runs the steps above to create a database based on new datasets specified by the user. Alternatively, you can complete `etl_pipeline.py` in the classroom on the `Project Workspace IDE` coming later.